src/systems/userinput/userinput.js

```diff
@@ -22,36 +21,113 @@ import { viveUserBindings } from "./bindings/vive-user";
 import { xboxControllerUserBindings } from "./bindings/xbox-controller-user";
 import { daydreamUserBindings } from "./bindings/daydream-user";
 
-import { updateActionSetsBasedOnSuperhands } from "./resolve-action-sets";
+import { resolveActionSets } from "./resolve-action-sets";
 import { GamepadDevice } from "./devices/gamepad";
 import { gamepadBindings } from "./bindings/generic-gamepad";
 
-const priorityMap = new Map();
-function prioritizeBindings(registeredMappings, activeSets) {
-  const activeBindings = new Set();
-  priorityMap.clear();
-  for (const mapping of registeredMappings) {
+const satisfiesPath = (binding, path) => {
+  return Object.values(binding.dest) && Object.values(binding.dest).indexOf(path) !== -1;
+};
+
+const satisfyPath = (bindings, path) => {
+  for (const binding of bindings) {
+    if (satisfiesPath(binding, path)) {
+      return true;
+    }
+  }
+  return false;
+};
+
+const satisfiedBy = (binding, bindings) => {
+  for (const path of Object.values(binding.src)) {
+    if (path.startsWith("/device/")) continue;
+    if (!satisfyPath(bindings, path)) return false;
+  }
+  return true;
+};
+
+function dependencySort(mappings) {
+  const unsorted = [];
+  for (const mapping of mappings) {
     for (const setName in mapping) {
-      if (!activeSets.has(setName) || !mapping[setName]) continue;
       for (const binding of mapping[setName]) {
-        const { root, priority } = binding;
-        const prevBinding = priorityMap.get(root);
-        if (!root || !priority) {
-          activeBindings.add(binding);
-        } else if (!prevBinding) {
-          activeBindings.add(binding);
-          priorityMap.set(root, binding);
-        } else if (priority > prevBinding.priority) {
-          activeBindings.delete(priorityMap.get(root));
-          activeBindings.add(binding);
-          priorityMap.set(root, binding);
-        } else if (prevBinding.priority === priority) {
-          console.error("equal priorities on same root", binding, priorityMap.get(root));
-        }
+        unsorted.push(binding);
+      }
+    }
+  }
+
+  const sorted = [];
+  while (unsorted.length > 0) {
+    const binding = unsorted.shift();
+    if (satisfiedBy(binding, sorted)) {
+      sorted.push(binding);
+    } else {
+      unsorted.push(binding);
```

Comment:
I guess you tested this and it turns out to be acceptably fast, but I'm honestly surprised that it is. Isn't this N^2 with respect to the number of bindings? And there can be quite a fair number of bindings active, like maybe one or two hundred, right? And the work for each step is not completely trivial, it's making some garbage and doing some string compares and so on.

It seems like you could do a lot of different ways that were faster, but maybe the simplest would be to maintain a version of the bindings that were indexed by their destination paths. Then checking whether a binding was satisfied would be fast.

Reply:
I share your concern that this code is not fast enough, and think that basically anything that takes any significant chunk of time can be made faster (with just a few exceptions like when we perform a raycast). The userinput system doesn't need to do a lot of work. That said, I was hesitant to start trying to remove some of these N^2 loops without first trying to measure how long the system tick was actually taking, whether the hotspots are where we think, and whether my changes actually sped the system up. For instance, this code is only run when active devices change, which I expect is very infrequent. This _can_ be made faster, but it doesn't matter to me as much as getting e.g. `computeExecutionStrategy` to be faster because it runs every time action sets change (very frequent).

To that end, I took several measurements using perf.html and focused on worst offenders / lowest hanging fruit. I *think* I managed to drop the average CPU time of the userinput system in half, from 4.6+-1% of the work done by javascript each frame to 2.3 +- 0.5%. Below I'll post my notes, the perf.html recordings, and the changes I made to help speed things up.

Reply:
First perf:
https://perfht.ml/2B2U5nE
System takes 4.6% total.
Resolve action sets: 18.694362017% : 18.9 / 101.1 
Compute execution strategy: 14.638971315% : 14.8 / 101.1

After removal of querySelectors from resolveActionSets:
https://perfht.ml/2B5dILP
System takes 4.2%
Resolve action sets        : 14.209320091% : 18.6 / 130.9
Compute execution strategy : 14.820473644% : 19.4/130.9

After replacing `for in` with `for of` and turn `masks` into an array of arrays, not a square matrix in a single array:
https://perfht.ml/2B2xSG7
System takes 2.7%
Resolve action sets        : 16.355140186% : 17.5 / 107
Compute execution strategy : 12.429906542%  : 13.3/107.0

After the next chunk of perf improvements:
https://perfht.ml/2B379JI
System takes 2.3%
Resolve action sets: 17.477477477% : 19.4/111.0
Compute execution strategy : 5.76576576577% : 6.4 / 111.0

Reply:
https://github.com/mozilla/hubs/pull/663/commits/c7eaf2ac36ba18560814f583b3f0481ec59b4ac2

Reply:
To be clear, I'm not content with the fact that the userinput system tick is still taking ~2.3% of the frame time, because I think we can get it to ~1% (general estimate, thinking that the raycasting that will definitely happen when you have a mouse takes about ~0.7% of the frametime). However, the userinput system in production now takes more like ~3.5% of the frame, I'm OK with pushing this as-is and making it faster as we move forward. There is other low-hanging perf fruit that we should be able to address when we prioritize them. For example, we could remove tracked-controls or superhands, since the event emitting that happens in both of those takes a surprisingly large amount of the CPU.